app/models/accounting/transaction.rb

```diff
@@ -95,4 +112,27 @@ def lookup_currency_id
       project.currency_id
     end
   end
+
+  def calculate_loan_amounts
+    loan_id = self.project_id
+
+    if self.loan_transaction_type == 'disbursement'
+      self.update_column(:change_in_principal, amount)
+    end
+
+    if amount <= interest_balance
+      self.change_in_interest =- amount
+      self.update_column(:change_in_interest, change_in_interest)
+    elsif amount > interest_balance
+      self.change_in_interest =- interest_balance
+      self.update_column(:change_in_interest, change_in_interest)
+      self.change_in_principal =- (interest_balance - amount)
+      self.update_column(:change_in_principal, change_in_principal)
+    end
+
+    if self.class.previous(loan_id)
+      self.interest_balance = self.class.previous(loan_id).interest_balance + change_in_interest
+      self.update_column(:interest_balance, interest_balance)
```

Comment:
Don't forget to update the `principal_balance` as well. (and recall comments from above about using `self` and removing all these excess `update_column` calls. You should just be able to assign all of these values and then `save!`